### PR TITLE
Add ability to parse and apply advanced resource filtering.

### PR DESCRIPTION
**Pull request by maruqu**
This change is [Reviewable](https://reviewable.io/reviews/socialwifi/flask-jsonapi/24)

### CODE REVIEW

**Comment by jgoclawski** [comment]
Reviewed 11 of 11 files at r1, 4 of 6 files at r2.
Review status: 9 of 19 files reviewed, 1 unresolved discussion (waiting on @jgoclawski and @maruqu)

---

*[tests/sqlalchemy_repository/conftest.py, line 38 at r1](https://reviewable.io/reviews/socialwifi/flask-jsonapi/24#-LGArlc12XDO8ZycAzuc:-LGArlc12XDO8ZycAzud:ba7x03) ([raw file](https://github.com/socialwifi/flask-jsonapi/blob/57a2a47ad09ee2cbc306263dcf7565e4f112b2dc/tests/sqlalchemy_repository/conftest.py#L38)):*
> ```Python
> def setup_db_schema(request, db_engine, db_session):
>     model_base = request.param
>     model_base.query = db_session.query_property(sqlalchemy_django_query.DjangoQuery)
> ```

As discussed, let's not force users to use DjangoQuery for their models.

---


*Comments from [Reviewable](https://reviewable.io/reviews/socialwifi/flask-jsonapi/24)*

File: flask_jsonapi/filters_schema.py [hunk 1]
```diff
@@ -1,29 +1,72 @@
...
+    YEAR = 'year'
+    MONTH = 'month'
+    DAY = 'day'
 
 
 class FilterField(utils.EqualityMixin):
-    def __init__(self, *, field_name=None, parse_value=str):
+    default_operator = Operators.EQ
+
+    def __init__(self, *, field_name=None, parse_value=str, operators=None, default_operator=None):
```
Comment (mimo00) [inline flask_jsonapi/filters_schema.py]: Why we need parse_value as param ?

File: flask_jsonapi/filters_schema.py [hunk 2]
```diff
@@ -37,3 +26,64 @@ def parse_value(self, value_string):
...
+        self.fields = getattr(meta, 'fields', ())
+
+
+class FilterSchemaMeta(type):
+    def __new__(cls, name, bases, attrs):
+        attrs['declared_filters'] = cls.get_declared_filters(bases, attrs)
+
+        new_class = super().__new__(cls, name, bases, attrs)
+        new_class._meta = FilterSchemaOptions(getattr(new_class, 'Meta', None))
+        new_class.base_filters = new_class.get_filters()
```
Comment (mimo00) [inline flask_jsonapi/filters_schema.py:105]: Is it necessary to set attribute of new_class that depends only on new_class method ? Is it responsibility of FilterSchemaMeta.
Then  code like:
filter_field = self.base_filters[filter_field_name]
Is kind of magic

File: flask_jsonapi/filters_schema.py [hunk 3]
```diff
@@ -84,6 +120,43 @@ def get_fields(cls):
...
+
+    @property
+    def request_filters(self) -> datastructures.MultiDict:
+        return datastructures.MultiDict(
+            [(key, value) for key, value
+             in flask.request.args.items(multi=True)
+             if key.startswith('filter')]
+        )
+
+    def _process_filter(self, key, value):
```
Comment (mimo00) [inline flask_jsonapi/filters_schema.py:173]: _process_filter do at least 2 things : 
1. Chooses proper filter_field
2. Parse a  key.
Parsing key causes strange params like: remaining_filter_attributes. 
It is better to move all parsing to FilterField, then responsibility will be clear.

**Comment by mimo00** [comment]
LGTM

**Pull request merged**